Source/Urho3D/Scene/Node.cpp

```diff
@@ -79,126 +81,206 @@ void Node::RegisterObject(Context* context)
 
-        // Skip temporary components
-        if (component && component->IsTemporary())
-            return;
+        const unsigned oldComponentId = static_cast<unsigned>(componentPrefab->GetId());
+        Component* component = SafeCreateComponent(
+            componentPrefab->GetTypeName(), componentPrefab->GetTypeNameHash(), discardIds ? 0 : oldComponentId);
+
+        resolver.AddComponent(oldComponentId, component);
+        componentPrefab->Export(component, flags);
 
-        // Serialize component
-        if (ArchiveBlock componentBlock = archive.OpenSafeUnorderedBlock("component"))
+        if (loadAsTemporary)
+            component->SetTemporary(true);
+    }
+
+    // Load children
+    const unsigned numChildren = reader.ReadNumChildren();
+    for (unsigned index = 0; index < numChildren; ++index)
+    {
+        reader.BeginChild();
         {
-            // Serialize component ID and type
-            unsigned componentID = component ? component->GetID() : 0;
-            StringHash componentType = component ? component->GetType() : StringHash::Empty;
-            const ea::string& componentTypeName = component ? component->GetTypeName() : EMPTY_STRING;
-            SerializeValue(archive, "id", componentID);
-            SerializeStringHash(archive, "type", componentType, componentTypeName);
-
-            // Create component if loading
-            if (loading)
-            {
-                component = SafeCreateComponent(EMPTY_STRING, componentType, componentID);
+            const SerializablePrefab* childPrefab = reader.ReadNode();
+            if (!childPrefab)
+                throw ArchiveException("Failed to read child prefab");
 
-                // Add component to resolver
-                resolver->AddComponent(componentID, component);
-            }
+            const unsigned oldChildId = static_cast<unsigned>(childPrefab->GetId());
+            Node* child = CreateChild(discardIds ? 0 : oldChildId);
+
+            const PrefabLoadFlags childFlags =
+                flags & ~PrefabLoadFlag::LoadAsTemporary & ~PrefabLoadFlag::IgnoreRootAttributes;
+            child->LoadInternal(*childPrefab, reader, resolver, childFlags);
 
-            // Serialize component.
-            component->SerializeInBlock(archive);
+            if (loadAsTemporary)
+                child->SetTemporary(true);
         }
-    });
+        reader.EndChild();
+    }
+}
 
-    // Skip children
-    if (!serializeChildren)
-        return;
+bool Node::Load(PrefabReader& reader, PrefabLoadFlags flags)
+{
+    try
+    {
+        const SerializablePrefab* nodePrefab = reader.ReadNode();
+        if (!nodePrefab)
+            throw ArchiveException("Failed to read node prefab");
+
+        SceneResolver resolver;
+        LoadInternal(*nodePrefab, reader, resolver, flags);
+
+        // Resolve IDs and apply attributes
+        resolver.Resolve();
+        ApplyAttributes();
```

Comment:
It seems that attributes applied to all children, even not temporal. Would it have unexpected side effects? Also is it going to call ApplyAttributes on PrefabReference component itself?

Reply:
Also does it matter in what order ApplyAttribute is called (i.e. top-to-bottom or bottom-to-top)?

Reply:
If it has unexpected side effects, then you wrote ApplyAttributes wrong.
It applies from root to children.

Reply:
Ok, but what about calling it on all components instead of only newly created ones?

Reply:
> Ok, but what about calling it on all components instead of only newly created ones?

What about it?

Reply:
Should it be avoided?

Reply:
It can be avoided, but why bother?
ApplyAttributes should be callable at any point of Component lifetime.

Reply:
> ApplyAttributes should be callable at any point of Component lifetime.
Is it mentioned in code comments or documentation?

/// Apply attribute changes that can not be applied immediately. Called after scene load or a network update.
This should probably mention that "ApplyAttributes should be callable at any point of Component lifetime."

Reply:
it seems that the existing ApplyAttributes  is well guarded behind flags (at least in cases I've looked into), so I'm just worrying someone may shoot them self in the foot.